plugins/cilium-cni/cilium-cni.go

```diff
@@ -238,75 +284,83 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return nil
 	})
 
-	req := ipam.IPAMReq{}
-	ipamConf, err := c.AllocateIP(ipam.CNIIPAMType, req)
+	ipam, err := client.IPAMAllocate("")
 	if err != nil {
 		return err
 	}
+
+	ep.IPV6 = ipam.IPV6
+	ep.IPV4 = ipam.IPV4
+
+	// release addresses on failure
 	defer func() {
-		if err != nil && ipamConf != nil {
-			if ipamConf.IP6 != nil {
-				req := ipam.IPAMReq{IP: &ipamConf.IP6.IP.IP}
-				if err = c.ReleaseIP(ipam.CNIIPAMType, req); err != nil {
-					log.Warningf("failed to release allocated IPv6 of container ID %q: %s", args.ContainerID, err)
-				}
-			}
-			if ipamConf.IP4 != nil {
-				req := ipam.IPAMReq{IP: &ipamConf.IP4.IP.IP}
-				if err = c.ReleaseIP(ipam.CNIIPAMType, req); err != nil {
-					log.Warningf("failed to release allocated IPv4 of container ID %q: %s", args.ContainerID, err)
-				}
+		if err != nil {
+			if err := releaseIPs(client, ep); err != nil {
+				log.Warningf("failed to cleanup IP allocation: %s", err)
 			}
 		}
 	}()
 
+	if err = plugins.SufficientAddressing(ipam.HostAddressing); err != nil {
+		return fmt.Errorf("%s", err)
+	}
+
+	state := CmdState{
+		Endpoint: ep,
+		Client:   client,
+		HostAddr: ipam.HostAddressing,
+	}
+
+	res := cniTypes.Result{}
+
+	if ipam.IPV6 != "" {
```

Comment:
We need an endpoint ID from ipam.IPV6, if it's empty, should we fail and return errors?